menpo3d/morphablemodel/fitter.py

```diff
@@ -0,0 +1,382 @@
+        instance = self._model.instance()
+
+        # Get view projection rotation matrices
+        view_t, proj_t, R = retrieve_view_projection_transforms(image, instance, group='ibug68')
+
+        # Get camera parameters array
+        rho_array = rho_from_view_projection_matrices(proj_t.h_matrix, R.h_matrix)
+
+        # Prior probabilities constants
+        alpha = np.zeros(n_alphas)
+        beta = np.zeros(n_betas)
+
+        dalpha_prior_dbeta = np.zeros(n_betas)
+        dalpha_prior_dalpha = 2. / np.square(self._model.shape_model.eigenvalues[:n_alphas])
+        
+        dbeta_prior_dalpha = np.zeros(n_alphas)
+        dbeta_prior_dbeta = 2. / np.square(self._model.texture_model.eigenvalues[:n_betas])
+        
+        if camera_update:
+            prior_drho = np.zeros(len(rho_array))
+            SD_alpha_prior = np.concatenate((dalpha_prior_dalpha, prior_drho, dalpha_prior_dbeta))
+            SD_beta_prior = np.concatenate((dbeta_prior_dalpha, prior_drho, dbeta_prior_dbeta))
+        else:
+            SD_alpha_prior = np.concatenate((dalpha_prior_dalpha, dalpha_prior_dbeta))
+            SD_beta_prior = np.concatenate((dbeta_prior_dalpha, dbeta_prior_dbeta))
+            
+        H_alpha_prior = np.diag(SD_alpha_prior)
+        H_beta_prior = np.diag(SD_beta_prior)
+        
+        # Initilialize rasterizer
+        rasterizer = GLRasterizer(height=image.height, width=image.width,
+                                  view_matrix=view_t.h_matrix,
+                                  projection_matrix=proj_t.h_matrix)  
+        errors = []
+        eps = np.inf
+        k = 0
+        
+        while k < max_iters and eps > threshold:
+            
+            # Progress bar
+            progress = k*100/max_iters
+            sys.stdout.write("\r%d%%" % progress)
+            sys.stdout.flush()
+
+            # Rotation matrices 
+            r_phi, r_theta, r_varphi = compute_rotation_matrices(rho_array)
+            
+            # Inverse rendering
+            tri_index_img, b_coords_img = rasterizer.rasterize_barycentric_coordinate_image(instance)
+            tri_indices = tri_index_img.as_vector() 
+            b_coords = b_coords_img.as_vector(keep_channels=True) 
+            yx = tri_index_img.mask.true_indices()
+
+            # Select triangles randomly
+            rand = np.random.permutation(b_coords.shape[1])
+            b_coords = b_coords[:, rand[:n_tris]]
+            yx = yx[rand[:n_tris]]
+            tri_indices = tri_indices[rand[:n_tris]]
+
+            # Build the vertex indices (3 per pixel)
+            # for the visible triangle
+            vertex_indices = instance.trilist[tri_indices]
+            
+            # Warp the shape witht the view matrix
+            W = view_t.apply(instance.points)
+            
+            # This solves the perspective projection problems
+            # It cancels the axes flip done in the view matrix before the rasterization
+            W[:, 1:] *= -1
```

Comment:
euuurgh this issue...
@patricksnape did we figure out in the end how to 'undo' the bad flip we had in the gold standard recovery code? Would be good to finally commit a correct version of that to menpo3d (that's what's getting included further up in this PR) but would obviously be preferable to have these issues fixed